### PR TITLE
Cache aware benchmark preview

### LINKED ISSUE #44341: Consider cache effects in Timer
## 🚀 Feature
When timing a statement, Timer should attempt to determine if the measurement is affected by CPU caching.

## Motivation

Consider the following measurements:
```
import torch
from torch.utils._benchmark import Timer

n = 1024
k = 2

m0 = Timer(
    "[torch.cat([x, y], 0, out=out) for _ in range(k)]",
    globals={"x": torch.ones((n, n)), "y": torch.ones((n, n)), "out": torch.empty((2 * n, n)), "k": k}
).blocked_autorange(min_run_time=5)

m1 = Timer(
    "[torch.cat([xi, yi], 0, out=outi) for xi, yi, outi in zip(x, y, out)]",
    globals={
        "x": [torch.ones((n, n)) for _ in range(k)],
        "y": [torch.ones((n, n)) for _ in range(k)],
        "out": [torch.empty((2 * n, n)) for _ in range(k)]
    }
).blocked_autorange(min_run_time=5)
    
print(m0, "\n")
print(m1)
```

```
<torch.utils._benchmark.utils.common.Measurement object at 0x7f0f1ca6e950>
[torch.cat([x, y], 0, out=out) for _ in range(k)]
  Median: 1.52 ms
  IQR:    0.10 ms (1.48 to 1.58)
  3229 measurements, 1 runs per measurement, 1 thread 

<torch.utils._benchmark.utils.common.Measurement object at 0x7f0f1ca6e8d0>
[torch.cat([xi, yi], 0, out=outi) for xi, yi, outi in zip(x, y, out)]
  Median: 3.07 ms
  IQR:    0.44 ms (2.87 to 3.31)
  1610 measurements, 1 runs per measurement, 1 thread
  WARNING: Interquartile range is 14.3% of the median measurement.
           This could indicate system fluctuation.
```

The second measurement is both slower and less stable; however it reflects a real use case where the Tensors are not already in CPU cache. In effect, we want to perform many independent measurements rather than many correlated ones.

## Pitch

There is some space for design here. Initial ideas are:
* Run through a buffer to evict entries from the cache between measurements.
* Clone inputs and replicate the semantics of the second example. (This would probably have to be opt in.)

It's also not clear if the Timer should try to actively correct for cached data, or just warn if it is detected.

### PR DESCRIPTION
**Pull request by bitfort**
Fixes #44341

import torch; from torch.utils._benchmark import Timer; Timer(stmt='torch.sum(x)', globals={'x': 
     torch.ones((1000,1000))}).adaptive_autorange()
     Median: 84.57 us
     IQR:    8.45 us (81.13 to 89.59)
     179 measurements, 10 runs per measurement, 1 thread
     WARNING: Runtime is impacted by CPU caching effects.
              CPU Caching speeds up by about 2.35x.

import torch; from torch.utils._benchmark import Timer; Timer(stmt='torch.ones((10,10))').adaptive_autorange()

     torch.ones((10,10))
     Median: 3.15 us
     IQR:    0.26 us (3.04 to 3.31)
     5 measurements, 1000 runs per measurement, 1 thread

import torch; from torch.utils._benchmark import Timer; Timer(stmt='torch.ones((500,500))').adaptive_autorange()

     torch.ones((500,500))
     Median: 39.71 us
     IQR:    0.56 us (39.24 to 39.80)
     4 measurements, 100 runs per measurement, 1 thread
    WARNING: Runtime is impacted by CPU caching effects.
              CPU Caching speeds up by about 10.66x.

import torch; from torch.utils._benchmark import Timer; Timer(stmt='sum(range(1000))').adaptive_autorange()
  
     sum(range(1000))
     Median: 16.93 us
     IQR:    0.11 us (16.88 to 16.99)
     7 measurements, 100 runs per measurement, 1 thread

import torch; from torch.utils._benchmark import Timer; Timer(stmt='sum(range(10000))').adaptive_autorange()

     sum(range(10000))
     Median: 191.01 us
     IQR:    5.11 us (189.92 to 195.03)
     6 measurements, 10 runs per measurement, 1 thread

### CODE REVIEW

File: torch/utils/_benchmark/utils/common.py [hunk 1]
```diff
@@ -21,6 +21,8 @@
 _IQR_WARN_THRESHOLD = 0.1
 _IQR_GROSS_WARN_THRESHOLD = 0.25
 
+_CACHE_SPEEDUP_THRESHOLD = 1.5
+_CACHE_SPEEDUP_MIN_SECONDS = 0.000005
```
Comment (robieta) [inline torch/utils/_benchmark/utils/common.py:25]: Can you either change this to `5e-6` or add a comment `# 5 us`?

File: torch/utils/_benchmark/utils/common.py [hunk 2]
```diff
@@ -50,6 +52,7 @@ def __init__(
         self.num_threads = num_threads
         self.stmt = stmt
         self.metadata = metadata
+        self._has_cache_warning = False
```
Comment (robieta) [inline torch/utils/_benchmark/utils/common.py:55]: I'm reluctant to formalize warnings too much right now, as it adds a lot of extra complexity. (For instance as is `_has_cache_warning` doesn't serialize/deserialize properly.) Let's just stick to adding an entry in `self._warnings`.

File: torch/utils/_benchmark/utils/timer.py [hunk 3]
```diff
@@ -105,26 +109,43 @@ def stop_hook(times):
...
+        number = 1
         with common.set_torch_threads(self._num_threads):
             # Estimate the block size needed for measurement to be negligible
             # compared to the inner loop. This also serves as a warmup.
-            overhead = np.median([self._timer.timeit(0) for _ in range(5)])
-            number = 1
             while True:
                 time_taken = self._timer.timeit(number)
+                if check_cache_speedup:
+                    cache_speedup = uncached_time / (time_taken / number)
```
Comment (robieta) [inline torch/utils/_benchmark/utils/timer.py:138]: This approach is not robust, as it implicitly assumes that measured time is linear with iterations. A better approach would be to run the early iterations both with and without a cache clear and compare the two. (In effect checking for an extra constant overhead.)

**Review by robieta — CHANGES_REQUESTED** [review]

File: torch/utils/_benchmark/utils/common.py (hunk 1 above)
Comment (bitfort) [reply torch/utils/_benchmark/utils/common.py:25]: "5e-6" sounds kind of scientific... How about 5 / (1000.0 * 1000)?

File: torch/utils/_benchmark/utils/common.py (hunk 2 above)
Comment (bitfort) [reply torch/utils/_benchmark/utils/common.py:55]: okay!

File: torch/utils/_benchmark/utils/timer.py (hunk 3 above)
Comment (bitfort) [reply torch/utils/_benchmark/utils/timer.py:138]: I don't think clearing the cache once then running 10 times is at all meaningful, the clear cache effects are reduced by an order of magnitude. Also, we only warn for a speedup of 1.5; so we would only detect for the N=10 case if the uncached slow down was 5x. Plus, the reported speedup would be misleading and possibly vary (depending on which check we hit it at).

We can compare 1-to-1 makes sense, but I suspect the clear isn't worth the added overhead for the 2nd round with N=10.

Comment (bitfort) [reply torch/utils/_benchmark/utils/timer.py:138]: Happy to run some experiments here if you'd like to dig into it.

**Pull request closed without merging**